### DISCUSSION TITLE
请教: 应该如何理解 adapter, driver, plugin 这些概念？

### DISCUSSION THREAD

**Discussion by luochen1990** [nonebot/discussions#482] [问答] [answered]
请教一下， nonebot 的 adapter，driver ，plugin 这些是怎么解耦的， [官方文档](https://nonebot.dev/docs/tutorial/fundamentals) 我没太看懂。

具体的问题有:

1. adapter 比较好理解， 就是适配不同的 IM 软件的， 但是 driver 这一层是为了屏蔽什么差异呢？
2. 如果我要开发一个聊天机器人，是要针对特定的 IM 软件去开发么（比如 tg 聊天机器人）？ 如果是这样那这几层解偶的意义何在？
3. 开发 plugin 需要假设 IM 么？ 还是说一个 plugin 应该能对任意 IM 都工作？
4. 如果我想开发一个功能，我是直接开发一个聊天机器人，还是先开发 plugin 然后在机器人里添加这个 plugin ？
5. 我怎么在python 包管理里添加 driver，adapter，plugin 依赖？ 比如 poetry add 或 pip install 的时候，要 add 什么名字？
6. 当我使用 onebot 启动机器人的时候，我是否被期望仅使用 CLI，而不用写任何 python 代码？
7. 我可以不使用 nb-cli 而仅使用传统的 python 编码的方式配置并启动机器人么？

**Comment by yanyongyu** [comment, marked as answer]
针对你的问题：

1. driver 这一层目前主要是体现 client 和 server 的差异，这在不同连接方式的情况下会产生不同的部署方式、影响某些功能的可用性（例如挂载其他服务）。
2. 如果你的机器人不考虑跨平台（或仅在特定多个平台），那么你完全可以针对你设想的平台进行开发，nonebot 也提供了对对接平台的校验等响应过滤的方法（[重载](https://nonebot.dev/docs/appendices/overload)、[实践](https://nonebot.dev/docs/best-practice/multi-adapter)）。如果你的机器人考虑跨平台，那么你可以通过 [alconna 插件](https://nonebot.dev/docs/best-practice/alconna/)等方式做到一份代码在多平台运行。
3. 如果你的 plugin 仅自己使用那自然无需考虑是否支持多平台。如果你的插件想要分享到商店，在发布指南里提到了需要提供插件元数据，元数据内包含了插件支持的平台，用于商店展示。
4. 插件和机器人是不冲突的，机器人本身就是插件的集合。
5. 例如你要使用 fastapi 驱动器，那么你可以 `poetry add nonebot2[fastapi]`。如果要添加 adapter，那就是 `poetry add nonebot-adapter-xxx`。plugin 的包名由插件自己定义。以上这些包名你都可以在商店插件详情里看到。
6. 这两之间没有直接关系，如果你想要开发插件，那你就需要写代码（不代表不用 CLI）；如果你只想用商店的插件，那你可以只用 CLI。
7. 当然可以，文档的[这一部分](https://nonebot.dev/docs/tutorial/application)就是。

**Comment by luochen1990** [comment]
谢谢，我发现 zsh 里面需要用 `poetry add 'nonebot2[fastapi]'`, 不然方括号会被 zsh 理解为其他东西，这可能是我一开始安装失败的主要原因。

**Comment by luochen1990** [comment]
另外， 我发现用 `nb-cli init` 创建的目录里，会有 `pyproject.toml` 文件，并且里面有 `[tool.nonebot]` 部分，我想问如果不使用 `nb-cli` 的话，我需要手动在 `pyproject.toml` 中添加这个部分么？

**Reply by yanyongyu** [reply]
这一部分是给 nb-cli 读取用于启动的，如果你按照文档手动创建的项目，那么这一部分会写在入口的 python 脚本里，所以就不需要 pyproject 里的了。

**Comment by luochen1990** [comment]
再请教一个问题， 这个 `.env.prod` 文件里的内容，其实就是设置环境变量么？  我如果用 direnv 然后在 .envrc 文件里 export 环境变量是等效的么？

**Reply by yanyongyu** [reply]
`.env` 文件内的配置不仅仅是设置变量的功能，同时兼具声明配置的作用，nb 并不会主动去获取环境变量里的数值，只有在 `.env` 内存在的配置才会读取环境变量。同时，环境变量的优先级高于文件内的值。
